design/one-pager-controlled-rollout-of-composition-functions-v2.md

```diff
@@ -0,0 +1,198 @@
+that functions can be referenced by OCI reference rather than name:
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v1
+kind: Composition
+metadata:
+  name: example
+spec:
+  compositeTypeRef:
+    apiVersion: custom-api.example.org/v1alpha1
+    kind: AcmeBucket
+  mode: Pipeline
+  pipeline:
+  - step: patch-and-transform
+    functionRef:
+      package: xpkg.crossplane.io/crossplane-contrib/function-patch-and-transform:v0.8.2
+    input:
+      # Removed for brevity
+```
+
+The `package` field, if given, must contain a fully-qualified OCI reference
+(i.e., include a registry, repository, and tag or digest). It is invalid to
+provide both `name` and `package`.
+
+When `package` is provided, the composition revision controller will ensure that
+the specified function is running by creating a `FunctionRuntime` resource for
+it (see the package manager section below). If a `FunctionRuntime` already
+exists, the composition revision will be added as an owner reference.
+
+A user wishing to roll out a new version of `function-patch-and-transform`
+simply updates their composition to reference the new version, resulting in
+creation of a new composition revision and runtime creation for the new function
+version. The existing `compositionUpdatePolicy` and
+`compositionRevisionSelector` mechanisms on XRs can be used to progressively
+roll out the change:
+
+```yaml
+apiVersion: custom-api.example.org/v1alpha1
+kind: AcmeBucket
+metadata:
+  name: my-bucket
+spec:
+  compositionUpdatePolicy: Manual
+  compositionRevisionSelector:
+    matchLabels:
+      release-channel: alpha
+```
+
+In addition to allowing for progressive rollout, this new way of specifying
+functions is simpler and safer to use than the existing method. Users don't need
+to specify all their functions as dependencies or otherwise pre-install
+them. They also don't need to know how the package manager names dependency
+packages to construct the right `functionRef.name` in their composition. They
+are also guaranteed to get exactly the version they specify regardless of what
+other changes happen in the cluster; pacakges can be specified by digest for
+further safety.
+
+### Package Manager Changes
+
+Currently, both the package revision controller and the package runtime
+controller operate on package revision resources (`FunctionRevision` and
+`ProviderRevision`). The revision controller installs resources (CRDs, etc.)
+from the package, while the runtime controller creates runtime resources
+(deployments, services, etc.) for active revisions of packages that have a
+runtime component.
+
+In order to allow runtime-only packages, we will introduce new `FunctionRuntime`
+and `ProviderRuntime` resources, which will be reconciled by the package runtime
+controller. The revision controller will be changed to create a runtime resource
+corresponding to the active revision; the runtime controller will no longer
+operate on revisions directly.
```

Comment:
CC @turkenh, since you somewhat recently broke the runtime logic out into a separate controller.

@adamwg I don't think we need to block on feedback/review from Hasan, but it'd be good to get his thoughts on this approach if he has time.